pkg/utils/version_test.go

```diff
@@ -0,0 +1,146 @@
+package utils
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestParseVersion(t *testing.T) {
+	var tests = []struct {
+		input    string
+		expected Version
+	}{
+		{"1.2.3", Version{1, 2, 3, ""}},
+		{"v1.2.3", Version{1, 2, 3, ""}},
+		{"v1.2.3-foo", Version{1, 2, 3, "foo"}},
+		{"0.22.9993@bar-xyz", Version{0, 22, 9993, "bar-xyz"}},
+		{"x0.22.9993@bar-xyz", Version{}},
+		{"10.22+whatever", Version{10, 22, 0, "whatever"}},
+		{"10+whatever", Version{10, 0, 0, "whatever"}},
+		{"whatever-10-nonsense", Version{}},
+	}
+	for _, test := range tests {
+		if actual := ParseVersion(test.input); !reflect.DeepEqual(actual, test.expected) {
+			t.Errorf("Expected %q for %s, got %q", test.expected, test.input, actual)
+		}
+	}
+}
+
+func TestIsUndefined(t *testing.T) {
+	var tests = []struct {
+		input    string
+		expected bool
+	}{
+		{"1.2.3", false},
+		{"v1.2.3", false},
+		{"0.22.9993@bar-xyz", false},
+		{"x0.22.9993@bar-xyz", true},
+		{"10.22+whatever", false},
+		{"10+whatever", false},
+		{"whatever-10-nonsense", true},
+	}
+	for _, test := range tests {
+		v := ParseVersion(test.input)
+		if actual := v.IsUndefined(); actual != test.expected {
+			t.Errorf("Expected IsUndefined() for %s to be %v, got %v", test.input, test.expected, actual)
+		}
+	}
+}
+
+func TestEquivalent(t *testing.T) {
+	var tests = []struct {
+		a        string
+		b        string
+		expected bool
+	}{
+		{"1.2.3", "1.2.3", true},
+		{"1.2.3", "v1.2.3", true},
+		{"v1.2.3", "v1.2.3", true},
+		{"v1.2.3", "1.2.3", true},
+		{"v1.2.3", "x1.2.3", false},
+		{"1.2.3", "0.2.3", false},
+		{"1.2.3", "1.2.4", false},
+		{"1.2.3", "1.3.2", false},
+		{"0.22.9993@bar-xyz", "0.22.9993", true},
+		{"0.22.9993@bar-xyz", "0.22.9993+xyz", true},
+		{"0.22.9993@bar-xyz", "0.22.999@bar-xyz", false},
+		{"10.22+whatever", "10.22", true},
+		{"10.22+whatever", "10.22_ignoreme", true},
+		{"10.22+whatever", "10.32_ignoreme", false},
+		{"10+whatever", "10", true},
+		{"10+whatever", "10%rrr", true},
+		{"cat", "dog", true},
```

Comment:
Adding from comments above, should this be false either because version is not defined and/or qualifiers should be compare in numeric/ascii lexical order?

Reply:
I don't think it make sense to compare qualifiers in lexical order. As above I took the view that if you cannot determine ordering between two versions, they are essentially equivalent. Released version should never be undefined. So two released versions should always be comparable. A released version is always considered ahead of an undefined version.